adapters/repos/db/queue/queue.go

```diff
@@ -0,0 +1,1063 @@
+	err := os.MkdirAll(q.dir, 0o755)
+	if err != nil {
+		return errors.Wrap(err, "failed to create directory")
+	}
+
+	// create chunk reader
+	q.r, err = newChunkReader(q.dir)
+	if err != nil {
+		return errors.Wrap(err, "failed to create chunk reader")
+	}
+
+	// create chunk writer
+	q.w, err = newChunkWriter(q.dir, q.r, q.Logger, q.chunkSize)
+	if err != nil {
+		return errors.Wrap(err, "failed to create chunk writer")
+	}
+
+	// determine the number of records stored on disk
+	// and the disk usage
+	err = q.analyzeDisk()
+	if err != nil {
+		return err
+	}
+
+	// set the last push time to now
+	q.lastPushTime = time.Now()
+
+	return nil
+}
+
+// Close the queue, prevent further pushes and unregister it from the scheduler.
+func (q *DiskQueue) Close() error {
+	if q == nil {
+		return nil
+	}
+
+	q.m.Lock()
+	if q.closed {
+		q.m.Unlock()
+		return errors.New("queue already closed")
+	}
+	q.closed = true
+	q.m.Unlock()
+
+	q.scheduler.UnregisterQueue(q.id)
+
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	if q.w != nil {
+		err := q.w.Close()
+		if err != nil {
+			return errors.Wrap(err, "failed to close chunk writer")
+		}
+	}
+
+	if q.r != nil {
+		err := q.r.Close()
+		if err != nil {
+			return errors.Wrap(err, "failed to close chunk reader")
+		}
+	}
+
+	return nil
+}
+
+func (q *DiskQueue) Metrics() *Metrics {
+	return q.metrics
+}
+
+func (q *DiskQueue) ID() string {
+	return q.id
+}
+
+var bufPool = sync.Pool{
+	New: func() any {
+		return new(bytes.Buffer)
+	},
+}
+
+func (q *DiskQueue) Push(record []byte) error {
+	q.m.RLock()
+	if q.closed {
+		q.m.RUnlock()
+		return errors.New("queue closed")
+	}
+	q.m.RUnlock()
+
+	if len(record) == 0 {
+		return errors.New("empty record")
+	}
+
+	buf := bufPool.Get().(*bytes.Buffer)
+	defer bufPool.Put(buf)
+
+	buf.Reset()
+
+	var bytesBuf [4]byte
+	// length of the record in 4 bytes
+	binary.BigEndian.PutUint32(bytesBuf[:], uint32(len(record)))
+	_, err := buf.Write(bytesBuf[:])
+	if err != nil {
+		return errors.Wrap(err, "failed to write record length")
+	}
+
+	// write the record
+	_, err = buf.Write(record)
+	if err != nil {
+		return errors.Wrap(err, "failed to write record")
+	}
+
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	q.lastPushTime = time.Now()
+
+	n, err := q.w.Write(buf.Bytes())
+	if err != nil {
+		return errors.Wrap(err, "failed to write record")
+	}
+
+	q.recordCount++
+	q.diskUsage += int64(n)
+
+	return nil
+}
+
+func (q *DiskQueue) Scheduler() *Scheduler {
+	return q.scheduler
+}
+
+func (q *DiskQueue) Flush() error {
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	return q.w.Flush()
+}
+
+func (q *DiskQueue) DequeueBatch() (batch *Batch, err error) {
+	c, err := q.r.ReadChunk()
+	if err != nil {
+		return nil, err
+	}
+
+	// if there are no more chunks to read,
+	// check if the partial chunk is stale (e.g no tasks were pushed for a while)
+	if c == nil || c.f == nil {
+		c, err = q.checkIfStale()
+		if c == nil || err != nil || c.f == nil {
+			return nil, err
```

Comment:
hmm, so here we could potencially have a situation where `c == nil` and `err == nil` then we would return `nil, nil`? is that expected?

Reply:
yeah both case are expected 👍🏼